localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp

```diff
@@ -0,0 +1,233 @@
+// Copyright 2023 Autoware Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ndt_scan_matcher/tree_structured_parzen_estimator.hpp"
+
+#include <algorithm>
+#include <iostream>
+
+TreeStructuredParzenEstimator::TreeStructuredParzenEstimator(
+  const int64_t n_startup_trials, const double x_stddev, const double y_stddev,
+  const double z_stddev, const double roll_stddev, const double pitch_stddev)
+: engine_(std::random_device{}()),
+  dist_uniform_(-M_PI, M_PI),
+  dist_normal_(0.0, 1.0),
+  good_num_(0),
+  n_startup_trials_(n_startup_trials),
+  x_stddev_(x_stddev),
+  y_stddev_(y_stddev),
+  z_stddev_(z_stddev),
+  roll_stddev_(roll_stddev),
+  pitch_stddev_(pitch_stddev),
+  base_stddev_({x_stddev_, y_stddev_, z_stddev_, roll_stddev_, pitch_stddev_, yaw_stddev_})
+{
+}
+
+void TreeStructuredParzenEstimator::add_trial(const Trial & trial)
+{
+  trials_.push_back(trial);
+  std::sort(trials_.begin(), trials_.end(), [](const Trial & lhs, const Trial & rhs) {
+    return lhs.score > rhs.score;
+  });
+  good_num_ = std::count_if(trials_.begin(), trials_.end(), [this](const Trial & trial) {
+    return trial.score > MIN_GOOD_SCORE;
+  });
+  good_num_ = std::min(good_num_, static_cast<int64_t>(trials_.size() * MAX_GOOD_RATE));
+}
+
+TreeStructuredParzenEstimator::Input TreeStructuredParzenEstimator::get_next_input()
```

Comment:
Please add const
```suggestion
TreeStructuredParzenEstimator::Input TreeStructuredParzenEstimator::get_next_input() const
```

Reply:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/d0590cf656ae14e6c5abf25cb4e2aa109d2aad9d